From 9bc092d55a76ec5a647494223147dfc1f1700853 Mon Sep 17 00:00:00 2001
From: Christian Hewitt <christianshewitt@gmail.com>
Date: Sun, 30 Aug 2026 12:28:15 +0000
Subject: [PATCH 152/247] FIXUP: wifi: aic: drop 8800DC firmware loaders from
 the BSP

The BSP Makefile builds only the D80 U02 subset, so aic8800dc_compat.o
is not in aic8800_bsp-y. aic_bsp_driver.c still carried the 8800DC
firmware loading helpers, and aicwf_plat_patch_load_8800dc() reads
chip_sub_id, which is defined in the unbuilt aic8800dc_compat.c. With
CONFIG_AIC8800_BSP_SUPPORT=y the reference survives to the vmlinux
link and breaks it:

  ld: aic_bsp_driver.o: in function `aicwf_plat_patch_load_8800dc':
  aic_bsp_driver.c:(.text.unlikely+0x540): undefined reference to
    `chip_sub_id'
  ld: aic_bsp_driver.o: relocation R_AARCH64_ADR_PREL_PG_HI21 against
    symbol `chip_sub_id' which may bind externally can not be used when
    making a shared object; recompile with -fPIC

The relocation diagnostic comes from arm64 building the kernel PIE; the
undefined symbol is the whole of the problem.

Remove the DC-only helpers rather than restoring the object they need.
Every caller of aicwf_patch_table_load(), aicwf_plat_patch_load_8800dc(),
aicwf_plat_rftest_load_8800dc(), aicwf_misc_ram_valid_check_8800dc() and
aicwf_plat_calib_load_8800dc() lives in aic8800dc_compat.c, so none of
them are reachable in the D80 U02 configuration this series targets.
The D80 firmware path is untouched.

Fixes: b5950965f075 ("FROMLIST(v4): wifi: aic: add SDIO BSP and build configuration")
Assisted-by: Claude:claude-opus-5
Signed-off-by: Christian Hewitt <christianshewitt@gmail.com>
---
 .../wireless/aic/aic8800_bsp/aic_bsp_driver.c | 202 ------------------
 .../wireless/aic/aic8800_bsp/aic_bsp_driver.h |   9 -
 2 files changed, 211 deletions(-)

diff --git a/drivers/net/wireless/aic/aic8800_bsp/aic_bsp_driver.c b/drivers/net/wireless/aic/aic8800_bsp/aic_bsp_driver.c
index 167c161b097d..b34b7ccbd0ce 100644
--- a/drivers/net/wireless/aic/aic8800_bsp/aic_bsp_driver.c
+++ b/drivers/net/wireless/aic/aic8800_bsp/aic_bsp_driver.c
@@ -20,7 +20,6 @@
 #include <linux/vmalloc.h>
 
 #include "aic8800d80_compat.h"
-#include "aic8800dc_compat.h"
 #include "aic_bsp_driver.h"
 #include "aicsdio.h"
 #include "aicsdio_txrxif.h"
@@ -798,207 +797,6 @@ int rwnx_plat_m2d_flash_ota_check(struct aic_sdio_dev *sdiodev, char *filename)
 }
 #endif // CONFIG_M2D_OTA_AUTO_SUPPORT
 
-int aicwf_patch_table_load(struct aic_sdio_dev *rwnx_hw, char *filename)
-{
-	struct device *dev = rwnx_hw->dev;
-	int err = 0;
-	unsigned int i = 0, size;
-	u32 *dst = NULL;
-	u8 *describle;
-	u32 fmacfw_patch_tbl_8800dc_u02_describe_size = 124;
-	u32 fmacfw_patch_tbl_8800dc_u02_describe_base; // read from patch_tbl
-
-	/* Copy the file on the Embedded side */
-	pr_info("aicbsp ### Upload %s\n", filename);
-
-	size = rwnx_load_firmware(&dst, filename, dev);
-	if (!dst) {
-		pr_err("aicbsp No such file or directory\n");
-		return -1;
-	}
-	if (size <= 0) {
-		pr_err("aicbsp wrong size of firmware file\n");
-		dst = NULL;
-		err = -1;
-	}
-
-	pr_info("aicbsp tbl size = %d\n", size);
-
-	fmacfw_patch_tbl_8800dc_u02_describe_base = dst[0];
-	AICWFDBG(LOGINFO, "FMACFW_PATCH_TBL_8800DC_U02_DESCRIBE_BASE = %x\n",
-		 fmacfw_patch_tbl_8800dc_u02_describe_base);
-
-	if (!err && i < size) {
-		err = rwnx_send_dbg_mem_block_write_req(rwnx_hw,
-							fmacfw_patch_tbl_8800dc_u02_describe_base,
-			fmacfw_patch_tbl_8800dc_u02_describe_size + 4, dst);
-		if (err)
-			pr_err("aicbsp write describe information fail\n");
-
-		describle =
-			kzalloc(fmacfw_patch_tbl_8800dc_u02_describe_size, GFP_KERNEL);
-		memcpy(describle, &dst[1], fmacfw_patch_tbl_8800dc_u02_describe_size);
-		pr_info("aicbsp %s", describle);
-		kfree(describle);
-		describle = NULL;
-	}
-
-	if (!err && i < size) {
-		for (i = (128 / 4); i < (size / 4); i += 2) {
-			pr_info("aicbsp patch_tbl:  %x  %x\n", dst[i], dst[i + 1]);
-			err = rwnx_send_dbg_mem_write_req(rwnx_hw, dst[i], dst[i + 1]);
-		}
-		if (err)
-			pr_err("aicbsp bin upload fail: %x, err:%d\r\n", dst[i], err);
-	}
-
-	if (dst) {
-		vfree(dst);
-		dst = NULL;
-	}
-
-	return err;
-}
-
-int aicwf_plat_patch_load_8800dc(struct aic_sdio_dev *sdiodev)
-{
-	int ret = 0;
-#if !defined(CONFIG_FPGA_VERIFICATION)
-	if (chip_sub_id == 0) {
-		pr_info("aicbsp u01 is loaing ###############\n");
-		ret = rwnx_plat_bin_fw_upload(sdiodev, ROM_FMAC_PATCH_ADDR,
-					      RWNX_MAC_PATCH_NAME2_8800DC);
-	} else if (chip_sub_id == 1) {
-		pr_info("aicbsp u02 is loaing ###############\n");
-		ret = rwnx_plat_bin_fw_upload(sdiodev, ROM_FMAC_PATCH_ADDR,
-					      RWNX_MAC_PATCH_NAME2_8800DC_U02);
-	} else if (chip_sub_id == 2) {
-		pr_info("aicbsp h_u02 is loaing ###############\n");
-		ret = rwnx_plat_bin_fw_upload(sdiodev, ROM_FMAC_PATCH_ADDR,
-					      RWNX_MAC_PATCH_NAME2_8800DC_H_U02);
-	} else {
-		pr_info("aicbsp unsupported id: %d\n", chip_sub_id);
-	}
-#endif
-	return ret;
-}
-
-int aicwf_plat_rftest_load_8800dc(struct aic_sdio_dev *sdiodev)
-{
-	int ret = 0;
-
-	ret = rwnx_plat_bin_fw_upload(sdiodev, RAM_LMAC_FW_ADDR,
-				      RWNX_MAC_FW_RF_BASE_NAME_8800DC);
-	if (ret) {
-		AICWFDBG(LOGINFO, "load rftest bin fail: %d\n", ret);
-		return ret;
-	}
-	return ret;
-}
-
-#ifdef CONFIG_DPD
-int aicwf_misc_ram_valid_check_8800dc(struct aic_sdio_dev *sdiodev,
-				      int *valid_out)
-{
-	int ret = 0;
-	u32 cfg_base = 0x10164;
-	struct dbg_mem_read_cfm cfm;
-	u32 misc_ram_addr;
-	u32 ram_base_addr, ram_word_cnt;
-	u32 bit_mask[4];
-	int i;
-
-	if (valid_out)
-		*valid_out = 0;
-
-	if (testmode == FW_RFTEST_MODE) {
-		u32 vect1 = 0;
-		u32 vect2 = 0;
-
-		cfg_base = RAM_LMAC_FW_ADDR + 0x0004;
-		ret = rwnx_send_dbg_mem_read_req(sdiodev, cfg_base, &cfm);
-		if (ret) {
-			AICWFDBG(LOGERROR, "cfg_base:%x vcet1 rd fail: %d\n", cfg_base,
-				 ret);
-			return ret;
-		}
-		vect1 = cfm.memdata;
-		if ((vect1 & 0xFFFF0000) != (RAM_LMAC_FW_ADDR & 0xFFFF0000)) {
-			AICWFDBG(LOGERROR, "vect1 invalid: %x\n", vect1);
-			return ret;
-		}
-		cfg_base = RAM_LMAC_FW_ADDR + 0x0008;
-		ret = rwnx_send_dbg_mem_read_req(sdiodev, cfg_base, &cfm);
-		if (ret) {
-			AICWFDBG(LOGERROR, "cfg_base:%x vcet2 rd fail: %d\n", cfg_base,
-				 ret);
-			return ret;
-		}
-		vect2 = cfm.memdata;
-		if ((vect2 & 0xFFFF0000) != (RAM_LMAC_FW_ADDR & 0xFFFF0000)) {
-			AICWFDBG(LOGERROR, "vect2 invalid: %x\n", vect2);
-			return ret;
-		}
-
-		cfg_base = RAM_LMAC_FW_ADDR + 0x0164;
-	}
-	// init misc ram
-	ret = rwnx_send_dbg_mem_read_req(sdiodev, cfg_base + 0x14, &cfm);
-	if (ret) {
-		AICWFDBG(LOGERROR, "rf misc ram[0x%x] rd fail: %d\n", cfg_base + 0x14,
-			 ret);
-		return ret;
-	}
-	misc_ram_addr = cfm.memdata;
-	AICWFDBG(LOGERROR, "misc_ram_addr=%x\n", misc_ram_addr);
-	// bit_mask
-	ram_base_addr = misc_ram_addr + offsetof(struct rf_misc_ram_t, bit_mask);
-	ram_word_cnt = (MEMBER_SIZE(struct rf_misc_ram_t, bit_mask) +
-					MEMBER_SIZE(struct rf_misc_ram_t, reserved)) /
-				   4;
-	for (i = 0; i < ram_word_cnt; i++) {
-		ret = rwnx_send_dbg_mem_read_req(sdiodev, ram_base_addr + i * 4, &cfm);
-		if (ret) {
-			AICWFDBG(LOGERROR, "bit_mask[0x%x] rd fail: %d\n",
-				 ram_base_addr + i * 4, ret);
-			return ret;
-		}
-		bit_mask[i] = cfm.memdata;
-	}
-	AICWFDBG(LOGTRACE, "bit_mask:%x,%x,%x,%x\n", bit_mask[0], bit_mask[1],
-		 bit_mask[2], bit_mask[3]);
-	if (bit_mask[0] == 0 && ((bit_mask[1] & 0xFFF00000) == 0x80000000) &&
-	    bit_mask[2] == 0 && ((bit_mask[3] & 0xFFFFFF00) == 0x00000000)) {
-		if (valid_out)
-			*valid_out = 1;
-	}
-	return ret;
-}
-
-int aicwf_plat_calib_load_8800dc(struct aic_sdio_dev *sdiodev)
-{
-	int ret = 0;
-
-	if (chip_sub_id == 1) {
-		ret = rwnx_plat_bin_fw_upload(sdiodev, ROM_FMAC_CALIB_ADDR,
-					      RWNX_MAC_CALIB_NAME_8800DC_U02);
-		if (ret) {
-			AICWFDBG(LOGINFO, "load rftest bin fail: %d\n", ret);
-			return ret;
-		}
-	} else if (chip_sub_id == 2) {
-		ret = rwnx_plat_bin_fw_upload(sdiodev, ROM_FMAC_CALIB_ADDR,
-					      RWNX_MAC_CALIB_NAME_8800DC_H_U02);
-		if (ret) {
-			AICWFDBG(LOGINFO, "load rftest bin fail: %d\n", ret);
-			return ret;
-		}
-	}
-	return ret;
-}
-
-#endif
-
 #ifdef CONFIG_DPD
 /**
  * var dpd_res - DPD calibration result shared with the WLAN driver
diff --git a/drivers/net/wireless/aic/aic8800_bsp/aic_bsp_driver.h b/drivers/net/wireless/aic/aic8800_bsp/aic_bsp_driver.h
index 16da7a255722..5c6371f2a94e 100644
--- a/drivers/net/wireless/aic/aic8800_bsp/aic_bsp_driver.h
+++ b/drivers/net/wireless/aic/aic8800_bsp/aic_bsp_driver.h
@@ -317,16 +317,7 @@ struct dbg_start_app_cfm {
 	u32 bootstatus;
 };
 
-int aicwf_plat_patch_load_8800dc(struct aic_sdio_dev *sdiodev);
-int aicwf_plat_rftest_load_8800dc(struct aic_sdio_dev *sdiodev);
-#ifdef CONFIG_DPD
-int aicwf_misc_ram_valid_check_8800dc(struct aic_sdio_dev *sdiodev,
-				      int *valid_out);
-int aicwf_plat_calib_load_8800dc(struct aic_sdio_dev *sdiodev);
-#endif
-
 int rwnx_load_firmware(u32 **fw_buf, const char *name, struct device *device);
-int aicwf_patch_table_load(struct aic_sdio_dev *rwnx_hw, char *filename);
 
 int rwnx_send_dbg_mem_read_req(struct aic_sdio_dev *sdiodev, u32 mem_addr,
 			       struct dbg_mem_read_cfm *cfm);
-- 
2.43.0

